Red/black tree implementation

andyross opened the pull request:
Seems to work.  Offline testing with gcov shows full coverage.  I'm very happy with the memory overhead and general structure.  Code size is 2.5k on x86, which is more than I'd like but still reasonable for the regimes where this will be applied.

Commit:
lib: Red/Black balanced tree data structure

A balanced tree implementation for Zephyr as we grow into bigger
regimes where simpler data structures aren't appropriate.

This implements an intrusive balanced tree that guarantees O(log2(N))
runtime for all operations and amortized O(1) behavior for creation
and destruction of whole trees.  The algorithms and naming are
conventional per existing academic and didactic implementations, c.f.:

https://en.wikipedia.org/wiki/Red%E2%80%93black_tree

The implementation is size-optimized to prioritize runtime memory
usage.  The data structure is intrusive, which is to say the struct
rbnode handle is intended to be placed in a separate struct the same
way other such structures (e.g. Zephyr's dlist list) and requires no
data pointer to be stored in the node.  The color bit is unioned with
a pointer (fairly common for such libraries).  Most notably, there is
no "parent" pointer stored in the node, the upper structure of the
tree being generated dynamically via a stack as the tree is recursed.
So the overall memory overhead of a node is just two pointers,
identical with a doubly-linked list.

Code size above dlist is about 2-2.5k on most architectures, which is
significant by Zephyr standards but probably still worthwhile in many
situations.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests: rbtree test

Test for the rbtree balanced tree.

The theory here is that the test uses a PRNG to randomly build and
modify trees build from an array of node objects (whose comparison
operator is just their memory address).  It begins with tiny node
sets, so as to catch edge cases (e.g. nodes being the root, or a leaf,
or parent/child) in the balancing algorithm before increasing the size
and continuing.

At each iteration the resulting tree is walked, verifying that it
contains exactly the nodes that we expect it should, that they appear
in the correct sorted order in the walk, that their local structure
matches a correct binary tree, and that the red/black rules are
followed always:

* The root must be black
* Red nodes can have only black children
* The path from the root to any NULL child pointer in the tree must
  cross the same number of black nodes.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andrewboie commented on misc/rb.c:
```diff
@@ -0,0 +1,495 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
The direction from Anas is to put this stuff under lib/, misc/ is now deprecated.
The headers can still be in include/misc/ AFAIK

andrewboie commented on misc/rb.c:
```diff
@@ -0,0 +1,495 @@
+		return;
+	}
+
+	struct rbnode *stack[tree->max_depth + 1];
```
I think checkpatch may barf on this, the coding style really wants all declarations to precede all code

andyross replied:
checkpatch seems to be allowing that, actually.  Not sure if that's a bug or an intended feature, but local declarations IMHO help a lot with clarity so I've been sneaking these by.

lpereira replied:
Is there any way this can blow up the stack? This is supposed to be used in the scheduler, and it might be able to blow the kernel stack by creating a lot of threads

andyross replied:
max_depth is bounded to 2*log2(N) (it's a balanced tree), so not really.   Creating a billion threads would require 240 bytes of stack to traverse.

andrewboie commented on misc/rb.c:
```diff
@@ -0,0 +1,495 @@
+		/* Cases where the sibling has only black children
+		 * have simple resolutions
+		 */
+		c0 = get_child(sib, 0), c1 = get_child(sib, 1);
```
curious why the comma operator used instead of just putting each on a line?

andyross replied:
Because it was a cut and paste of a previous declaration and I didn't change the syntax completely.   Will fix.

lpereira commented on misc/rb.c:
```diff
@@ -0,0 +1,495 @@
+				 * has a missing black so iterate
+				 * upward
+				 */
+				set_color(sib, RED);
```
If you're coloring `sib` red, maybe if you call `set_color(sib, RED)` outside the `if`/`else`, just once, will save a few bytes

andyross replied:
Good point.  Though to be fair set_color() is pure, inline and available above.  I'd expect the compiler would have hoisted those instructions anyway.  Will fix.

tyrantlizard77 reviewed (commented):
looks very nice, just a few comments

lpereira commented on misc/rb.c:
```diff
@@ -0,0 +1,495 @@
+	if (side) {
+		return n->children[1];
+	} else {
+		long l = (long) n->children[0];
```
Use `uintptr_t` instead of `long`.

lpereira commented on misc/rb.c:
```diff
@@ -0,0 +1,495 @@
+		return n->children[1];
+	} else {
+		long l = (long) n->children[0];
+		l &= ~1;
```
`~1l`?

lpereira commented on misc/rb.c:
```diff
@@ -0,0 +1,495 @@
+		long old = (long) n->children[0];
+		long new = (long) val;
+
+		n->children[0] = (void *) (new | (old & 1));
```
`& ~1l`?

lpereira commented on tests/misc/rbtree/src/main.c:
```diff
@@ -0,0 +1,211 @@
+void set_node_mask(int node, int val)
+{
+	unsigned int *p = &node_mask[node / 32];
+	unsigned int bit = 1 << (node % 32);
```
This should be `1u << (node % 32)`: shifting left by 31 on an `int` is undefined behavior

lpereira commented on tests/misc/rbtree/src/main.c:
```diff
@@ -0,0 +1,211 @@
+int get_node_mask(int node)
+{
+	unsigned int *p = &node_mask[node / 32];
+	unsigned int bit = 1 << (node % 32);
```
Same here.

andyross commented:
Updated to address file location and bit twiddling typesafe pedantry.  Should be good to go.

andyross commented:
Rebased to fix collision in lib/CMakeLists.txt

andrewboie commented:
This needs a rebase

andrewboie commented:
> we have been putting things inside directories under lib/, even if it is a one file implementation

What's the rationale for doing that?
To me I'd rather see a "zephyr-runtime" library defined and all this miscellaneous stuff coalesced under that.

nashif commented:
> What's the rationale for doing that?
> To me I'd rather see a "zephyr-runtime" library defined and all this miscellaneous stuff coalesced under that.

a library can consist of one file, or 100 files and a very complex configuration infrastructure, just for the sake of keeping things organised and self-contained, i.e. keep it modular and keep parts of the libraries separated from each other and self-contained. This way for example you can easily remove, replace and substitute implementations, eventually integrate tests/documentation in the same folder and so on.

andrewboie commented:
Have you given any thought to how these are going to be organized into .a or .so files when we support non monlithic kernels? Dozens of tiny libraries that everyone has to pull in anyway? That sounds awful to deal with

Put minimal in its own lib, big stuff like crypto into their own libs, but having dozens of random tiny libraries that ought to be just put together into a runtime lib is insane to me

On Apr 6, 2018 2:59 PM, Anas Nashif <notifications@github.com> wrote:

What's the rationale for doing that?
To me I'd rather see a "zephyr-runtime" library defined and all this miscellaneous stuff coalesced under that.

a library can consist of one file, or 100 files and a very complex configuration infrastructure, just for the sake of keeping things organised and self-contained, i.e. keep it modular and keep parts of the libraries separated from each other and self-contained. This way for example you can easily remove, replace and substitute implementations, eventually integrate tests/documentation in the same folder and so on.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://github.com/zephyrproject-rtos/zephyr/pull/6738#issuecomment-379360873>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADOv6a1LMD0MIRrJwrMfzB7jJOFbE8A-ks5tl8jsgaJpZM4S10Qc>.

nashif commented:
> Put minimal in its own lib, big stuff like crypto into their own libs, but having dozens of random tiny libraries that ought to be just put together into a runtime lib is insane to me

I did not say we should do it this way.

All I am asking for is organising the file based on the features, how you build and archive them is something you can do regardless of the location of the file, rb.c or rbtree/rb.c

Commit:
lib/rbtree: Add a rb_contains() predicate

Returns true if the specified node is in the tree.  Allows the tree to
be used for "set" style semantics along with a lessthan_fn that simply
compares the nodes by their address.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
